File: ChuckNorrisIO/app/src/main/java/br/com/titoschmidt/chucknorrisio/presentation/JokePresenter.java
package br.com.titoschmidt.chucknorrisio.presentation;

import br.com.titoschmidt.chucknorrisio.JokeActivity;
import br.com.titoschmidt.chucknorrisio.datasource.JokeRemoteDataSource;
import br.com.titoschmidt.chucknorrisio.models.Joke;

public class JokePresenter implements JokeRemoteDataSource.JokeCallback {

    private final JokeActivity jokeActivity;
    private final JokeRemoteDataSource dataSource;

    public JokePresenter(JokeActivity jokeActivity, JokeRemoteDataSource dataSource) {
        this.jokeActivity = jokeActivity;
        this.dataSource = dataSource;
    }

    public void findJokeBy(String category){
        jokeActivity.mostraProgressBar();
        this.dataSource.findJokeBy(this, category);
    }

    @Override
    public void onSuccess(Joke joke){
        jokeActivity.mostraJoke(joke);
    }
    @Override
    public void onError(String msgErro){
        jokeActivity.mostraFalha(msgErro);
    }

    public void onComplete(){
        jokeActivity.escondeProgressBar();
    }
}

File: JokerApp/app/src/main/java/co/titoschmidt/tutorial/jokerappdev/presenter/JokePresenter.kt
package co.titoschmidt.tutorial.jokerappdev.presenter

import co.titoschmidt.tutorial.jokerappdev.data.JokeCallback
import co.titoschmidt.tutorial.jokerappdev.data.JokeRemoteDataSource
import co.titoschmidt.tutorial.jokerappdev.model.Joke
import co.titoschmidt.tutorial.jokerappdev.view.JokeFragment

/*
 * 13/12/2021
 * @author titoesa@gmail.com (Tito Schmidt).
 * 
 */
class JokePresenter(
    private val view: JokeFragment,
    private val jokeDataSource: JokeRemoteDataSource
    ) : JokeCallback {

    fun findBy(category: String) {
        view.showProgressBar()
        jokeDataSource.findBy(this, category)
    }

    override fun onSuccess(response: Joke) {
        view.showJoke(response)
    }

    override fun onFailure(response: String) {
        view.showFailure(response)
    }

    override fun onComplete() {
        view.hideProgressBar()
    }
}